src/ray/core_worker/test/direct_task_transport_test.cc

```diff
@@ -1545,19 +1553,205 @@ TEST(DirectTaskTransportTest, TestPipeliningNumberOfWorkersRequested) {
+  // worker)
+  std::vector<TaskSpecification> tasks_stolen;
+  for (int i = 0; i < 5; i++) {
+    tasks_stolen.push_back(BuildEmptyTaskSpec());
+  }
+  ASSERT_TRUE(worker_client->ReplyStealTasks(Status::OK(), tasks_stolen));
+  tasks_stolen.clear();
+  ASSERT_TRUE(tasks_stolen.empty());
+  ASSERT_EQ(raylet_client->num_workers_requested, 3);
+  ASSERT_EQ(raylet_client->num_workers_returned, 0);
+  ASSERT_EQ(raylet_client->num_workers_disconnected, 0);
+  ASSERT_EQ(task_finisher->num_tasks_complete, 10);
+  ASSERT_EQ(task_finisher->num_tasks_failed, 0);
+  ASSERT_EQ(raylet_client->num_leases_canceled, 0);
+  ASSERT_EQ(worker_client->callbacks.size(), 10);
+  ASSERT_EQ(worker_client->steal_callbacks.size(), 0);
+
+  // The second worker finishes its workload of 5 tasks and begins stealing from the first
+  // worker
+  for (int i = 1; i <= 5; i++) {
+    ASSERT_TRUE(worker_client->ReplyPushTask());
+  }
+  ASSERT_EQ(raylet_client->num_workers_requested, 3);
+  ASSERT_EQ(raylet_client->num_workers_returned, 0);
+  ASSERT_EQ(raylet_client->num_workers_disconnected, 0);
+  ASSERT_EQ(task_finisher->num_tasks_complete, 15);
+  ASSERT_EQ(task_finisher->num_tasks_failed, 0);
+  ASSERT_EQ(raylet_client->num_leases_canceled, 0);
+  ASSERT_EQ(worker_client->callbacks.size(), 5);
+  ASSERT_EQ(worker_client->steal_callbacks.size(), 1);
+  // The second worker steals floor(5/2)=2 tasks from the first worker
+  for (int i = 1; i <= 2; i++) {
+    ASSERT_TRUE(worker_client->ReplyPushTask(Status::OK(), false, true));
+  }
+  ASSERT_EQ(raylet_client->num_workers_requested, 3);
+  ASSERT_EQ(raylet_client->num_workers_returned, 0);
+  ASSERT_EQ(raylet_client->num_workers_disconnected, 0);
+  ASSERT_EQ(task_finisher->num_tasks_complete, 15);
+  ASSERT_EQ(task_finisher->num_tasks_failed, 0);
+  ASSERT_EQ(raylet_client->num_leases_canceled, 0);
+  ASSERT_EQ(worker_client->callbacks.size(), 3);
+  ASSERT_EQ(worker_client->steal_callbacks.size(), 1);
+  ASSERT_TRUE(tasks_stolen.empty());
+  for (int i = 0; i < 2; i++) {
+    tasks_stolen.push_back(BuildEmptyTaskSpec());
+  }
+  ASSERT_FALSE(tasks_stolen.empty());
+  ASSERT_TRUE(worker_client->ReplyStealTasks(Status::OK(), tasks_stolen));
+  tasks_stolen.clear();
+  ASSERT_TRUE(tasks_stolen.empty());
+  ASSERT_EQ(raylet_client->num_workers_requested, 3);
+  ASSERT_EQ(raylet_client->num_workers_returned, 0);
+  ASSERT_EQ(raylet_client->num_workers_disconnected, 0);
+  ASSERT_EQ(task_finisher->num_tasks_complete, 15);
+  ASSERT_EQ(task_finisher->num_tasks_failed, 0);
+  ASSERT_EQ(raylet_client->num_leases_canceled, 0);
+  ASSERT_EQ(worker_client->callbacks.size(), 5);
+  ASSERT_EQ(worker_client->steal_callbacks.size(), 0);
+
+  // The first worker executes the remaining 3 tasks (the ones not stolen) and returns
+  for (int i = 1; i <= 3; i++) {
+    ASSERT_TRUE(worker_client->ReplyPushTask());
+  }
+  ASSERT_EQ(raylet_client->num_workers_requested, 3);
```

Comment:
Why do the 2 workers get returned here? I thought one worker would still be kept since there is 1 extra task.

Reply:
this has to do with the fact that the `SchedulingKeyEntry::StealableTasks() `function (https://github.com/gabrieleoliaro/ray/blob/atomic_work_stealing/src/ray/core_worker/transport/direct_task_transport.h#L324) that we use to quickly determine if there are stealable tasks is fast but not necessarily 100% correct all the time. For instance, if there are 10 active workers, and 5 tasks, all of which are queued at one worker, it would be beneficial for the remaining 9 workers to steal from the one that is overloaded, but since the number of tasks in flight (5) is less than the number of active workers (10), this won't happen. In the specific case of the test, we have 2 tasks and two active workers, so `total_tasks_in_flight `is NOT > `active_workers.size()`, and thus `SchedulingKeyEntry::StealableTasks()` returns false, resulting in one of the two workers being returned early.

Reply:
A way to fix this would be to let SchedulingKeyEntry::StealableTasks() return true if the number of tasks in flight to ANY of the active workers exceeds the number of active workers. This would be more precise, although perhaps a bit slower

Reply:
Ah I see, thanks! I think this is fine for now, but can you add a TODO in the StealableTasks() definition to explain why it's not exact?